partners_api/taxi_engine.hpp

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "storage/index.hpp"
+
+#include "partners_api/taxi_base.hpp"
+
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+namespace taxi
+{
+using SuccessfulCallback =
+    std::function<void(ProvidersContainer const & products, uint64_t const requestId)>;
+
+using ErrorCallback = std::function<void(ErrorsContainer const & errors, uint64_t const requestId)>;
+
+/// This class is used to collect replies from all taxi apis and to call callback when all replies
+/// are collected. The methods are called in callbacks on different threads, so synchronization is
+/// required.
+class ResultMaker
+{
+public:
+  void Reset(uint64_t requestId, size_t requestsCount, SuccessfulCallback const & successCallback,
+             ErrorCallback const & errorCallback);
+  /// Reduces number of requests outstanding.
+  void DecrementRequestCount(uint64_t requestId);
+  /// Processes successful callback from taxi api.
+  void ProcessProducts(uint64_t requestId, Provider::Type type,
+                       std::vector<Product> const & products);
+  /// Processes error callback from taxi api.
+  void ProcessError(uint64_t requestId, Provider::Type type, ErrorCode code);
+  /// Checks number of requests outstanding and calls callback when number is equal to zero.
+  void MakeResult(uint64_t requestId) const;
+
+private:
+  void DecrementRequestCount();
+
+  mutable std::mutex m_mutex;
+  uint64_t m_requestId = 0;
+  SuccessfulCallback m_successCallback;
+  ErrorCallback m_errorCallback;
+
+  int8_t m_requestsCount = 0;
```

Comment:
Please, explain in the comment to the code why int8_t is chosen and what's the meaning of the field.